File: recency_cv/gate.py
"""Recency-reducibility drift gate: expanding-vs-sliding CV scheme decision.

The default time-series CV scheme is expanding (use ALL available history). This
gate switches to sliding ONLY when the train→val drift is genuinely
recency-REDUCIBLE — that is, when restricting train to its most recent block
measurably brings the covariate distribution closer to val. That is a different
and much stronger condition than "drift is large." The classic drift detectors —
an adversarial train-vs-val AUC, or a per-feature KS statistic — answer "are
train and val distinguishable?", and on a forecasting holdout that question is
already settled before any drift exists: val is the future window by
construction, so a time index, seasonal phase, or any slow trend makes train and
val trivially separable. Those detectors saturate near their maximum on every
forecasting set, whether or not recency would help, and a gate built on them
picks sliding almost always — discarding history for no measurable gain.

This gate instead asks the question sliding can actually act on: does the recent
slice of train look more like val than the full slice does? For each shared
numeric covariate it computes a standardized mean shift, |mu_train - mu_val| /
pooled_std, twice — full train vs val and the last RECENT_PERIODS ranks of train
vs val — and measures the per-feature improvement. It aggregates that into
frac_improved (the breadth of features that move closer) and rel (the relative
size of the average move). Because the metric is an effect size on the very
quantity sliding manipulates, it does NOT saturate on "val is the future": a set
whose drift is pure trend shows little recent-vs-full improvement and stays on
expanding, while only a set where recency genuinely closes the gap trips the
sliding branch. Engineered seasonality / time-index columns are excluded from
the scan precisely because they encode "val is the future" rather than reducible
drift, and the val TARGET is never read — only competition-provided val
covariates participate.
"""
from __future__ import annotations

from typing import Optional

import numpy as np
import pandas as pd

from .thresholds import (
    RECENT_PERIODS,
    DRIFT_REL_THRESHOLD,
    DRIFT_FRAC_IMPROVED_THRESHOLD,
    MIN_FEATURES_FOR_SLIDING,
    DRIFT_IMPROVE_PER_FEATURE_THRESHOLD,
)

__all__ = ["standardized_mean_shift", "drift_diagnostic", "decide_scheme"]

# Engineered seasonality / time-index columns are excluded from the drift scan:
# they are deterministic functions of period rank, so they separate train from
# val by construction and encode "val is the future" rather than reducible drift.
_WINDOW_FEATURES = {
    "period_id_ord", "period_id_trend", "period_id_of_cycle", "horizon",
    "period_id_sin", "period_id_cos", "period_id_sin2", "period_id_cos2",
    "period_id_quarter", "period_id_month", "month_of_year",
    "quarter_of_year", "is_quarter_start",
}
_NON_FEATURES = {"adversarial_weights"}


def standardized_mean_shift(a, b) -> float:
    """|mean(a) - mean(b)| / pooled_std. Effect size; 0 = identical; does NOT
    saturate the way an adversarial AUC does.

    Drops NaNs from each side. If either side has fewer than 5 valid values,
    returns nan. pooled_std = sqrt((var(a) + var(b)) / 2) using population
    variance (np.var, ddof=0); if pooled_std is 0 returns 0.0.
    """
    a = np.asarray(a, dtype=float)
    b = np.asarray(b, dtype=float)
    a = a[~np.isnan(a)]
    b = b[~np.isnan(b)]
    if len(a) < 5 or len(b) < 5:
        return float("nan")
    sp = float(np.sqrt((np.var(a) + np.var(b)) / 2.0))
    if sp == 0.0:
        return 0.0
    return float(abs(np.mean(a) - np.mean(b)) / sp)


def drift_diagnostic(
    train_df: pd.DataFrame,
    val_df: Optional[pd.DataFrame],
    time_col: str,
    *,
    recent_periods: int = RECENT_PERIODS,
    exclude=None,
) -> dict:
    """Recent-vs-full drift diagnostic.

    Returns a verdict dict carrying the summary stats and an ``ok`` flag. When
    ``ok`` is False, ``reason`` explains why the diagnostic could not run — the
    caller falls back to expanding in that case (NEVER to sliding on missing
    evidence).

    The recent rank is derived directly from ``train_df[time_col]`` via
    ``pd.to_numeric(..., errors="coerce")``. If no numeric rank can be derived,
    if the rank span is shorter than ``recent_periods``, or if there are no
    shared numeric non-excluded features, the diagnostic returns ``ok=False``.

    ``_WINDOW_FEATURES`` and ``{"adversarial_weights"}`` are excluded from the
    scan by default; ``exclude`` adds further column names to that set. The val
    TARGET is never read — only shared numeric covariates participate.
    """
    excluded = set(_WINDOW_FEATURES) | set(_NON_FEATURES)
    if exclude:
        excluded |= set(exclude)

    if val_df is None or train_df is None:
        return {"ok": False, "reason": "no val frame available"}
    if not time_col or time_col not in train_df.columns:
        return {"ok": False, "reason": f"time_col {time_col!r} not present in train frame"}

    common = [
        c for c in train_df.columns
        if c in val_df.columns
        and pd.api.types.is_numeric_dtype(train_df[c])
        and c not in excluded
    ]
    if not common:
        return {"ok": False, "reason": "no shared numeric non-window features between train and val"}

    tr_rank = pd.to_numeric(train_df[time_col], errors="coerce")
    tr_rank_valid = tr_rank.dropna()
    if tr_rank_valid.empty:
        return {"ok": False, "reason": "could not derive numeric rank for train rows"}

    rank_max = int(tr_rank_valid.max())
    rank_span = rank_max + 1
    if rank_span < recent_periods:
        return {"ok": False,
                "reason": f"train spans {rank_span} periods < RECENT_PERIODS={recent_periods}"}
    cutoff = rank_max - recent_periods + 1
    recent_mask = (tr_rank >= cutoff).fillna(False).to_numpy()

    rows = []
    for c in common:
        a_full = train_df[c].to_numpy(dtype=float, copy=False)
        a_recent = train_df.loc[recent_mask, c].to_numpy(dtype=float, copy=False)
        b_val = val_df[c].to_numpy(dtype=float, copy=False)
        d_full = standardized_mean_shift(a_full, b_val)
        d_recent = standardized_mean_shift(a_recent, b_val)
        if np.isnan(d_full) or np.isnan(d_recent):
            continue
        rows.append((c, d_full, d_recent, d_full - d_recent))

    if not rows:
        return {"ok": False, "reason": "no feature yielded a valid distance (too many NaNs?)"}

    mean_full = float(np.mean([r[1] for r in rows]))
    mean_recent = float(np.mean([r[2] for r in rows]))
    mean_impr = float(np.mean([r[3] for r in rows]))
    frac_improved = float(np.mean([r[3] > DRIFT_IMPROVE_PER_FEATURE_THRESHOLD for r in rows]))
    rel = float(mean_impr / mean_full) if mean_full > 0 else 0.0

    return {
        "ok":                  True,
        "n_features_scanned":  len(rows),
        "rank_max":            rank_max,
        "recent_periods":      recent_periods,
        "recent_cutoff_rank":  cutoff,
        "mean_dist_full":      mean_full,
        "mean_dist_recent":    mean_recent,
        "mean_improvement":    mean_impr,
        "rel":                 rel,
        "frac_improved":       frac_improved,
        "per_feature": [
            {"feature": c, "dist_full": float(d_full),
             "dist_recent": float(d_recent), "improvement": float(d_full - d_recent)}
            for (c, d_full, d_recent, _) in rows
        ],
    }


def decide_scheme(diagnostic: dict) -> dict:
    """Map a drift diagnostic onto the CV scheme.

    Expanding is the structural default. Sliding is chosen ONLY when ALL three
    affirmative conditions hold:
        frac_improved >= DRIFT_FRAC_IMPROVED_THRESHOLD  (primary gate)
        rel           >= DRIFT_REL_THRESHOLD            (secondary gate)
        n_features    >= MIN_FEATURES_FOR_SLIDING       (evidence-breadth floor)
    Any failure — including a non-runnable diagnostic — falls back to expanding.

    Returns ``{"scheme", "reason", "gates": {frac_ok, rel_ok, nfeat_ok}}``.
    """
    if not diagnostic.get("ok"):
        return {
            "scheme": "expanding",
            "reason": (
                f"expanding (fallback - drift diagnostic could not run: "
                f"{diagnostic.get('reason', 'unknown')})"
            ),
            "gates": {"frac_ok": False, "rel_ok": False, "nfeat_ok": False},
        }

    rel = diagnostic["rel"]
    frac = diagnostic["frac_improved"]
    n_feat = int(diagnostic.get("n_features_scanned") or 0)
    frac_ok = frac >= DRIFT_FRAC_IMPROVED_THRESHOLD
    rel_ok = rel >= DRIFT_REL_THRESHOLD
    nfeat_ok = n_feat >= MIN_FEATURES_FOR_SLIDING
    gates = {"frac_ok": frac_ok, "rel_ok": rel_ok, "nfeat_ok": nfeat_ok}

    if frac_ok and rel_ok and nfeat_ok:
        reason = (
            f"sliding (drift is recency-reducible AND evidence is broad enough: "
            f"frac_improved={frac:.3f} >= {DRIFT_FRAC_IMPROVED_THRESHOLD} "
            f"AND rel={rel:.3f} >= {DRIFT_REL_THRESHOLD} "
            f"AND n_features={n_feat} >= {MIN_FEATURES_FOR_SLIDING})"
        )
        return {"scheme": "sliding", "reason": reason, "gates": gates}

    fails = []
    if not nfeat_ok:
        fails.append(
            f"insufficient features to assess drift breadth "
            f"(n_features={n_feat} < {MIN_FEATURES_FOR_SLIDING})"
        )
    if not frac_ok:
        fails.append(
            f"frac_improved={frac:.3f} < {DRIFT_FRAC_IMPROVED_THRESHOLD} (primary gate)"
        )
    if not rel_ok:
        fails.append(
            f"rel={rel:.3f} < {DRIFT_REL_THRESHOLD} (secondary gate)"
        )
    reason = (
        "expanding (conservative default; sliding not affirmatively justified - "
        + "; ".join(fails)
        + ")"
    )
    return {"scheme": "expanding", "reason": reason, "gates": gates}

File: demo/gate.js
// Dependency-free ES-module port of recency_cv/gate.py.
//
// Mirrors the Python reference EXACTLY: population variance (ddof=0), the
// <5-valid-values -> NaN rule, a strict per-feature improvement > 0.05 (NOT
// >=), the three sliding gates frac >= 0.60 AND rel >= 0.25 AND n >= 12, and the
// same default exclude sets. fixture_inputs.json + expected.json (under parity/)
// are the source of truth; this file must reproduce them.
//
// JS works on arrays of plain row objects where Python uses DataFrames:
//   driftDiagnostic(trainRows, valRows, timeCol, { recentPeriods, exclude })

// ── calibrated thresholds (see recency_cv/thresholds.py) ──────────────────────
export const RECENT_PERIODS = 14;
export const DRIFT_REL_THRESHOLD = 0.25;             // secondary gate
export const DRIFT_FRAC_IMPROVED_THRESHOLD = 0.6;    // primary gate (str(0.60) === "0.6")
export const MIN_FEATURES_FOR_SLIDING = 12;          // evidence-breadth floor
export const DRIFT_IMPROVE_PER_FEATURE_THRESHOLD = 0.05;

// Engineered seasonality / time-index columns: excluded by construction.
const WINDOW_FEATURES = new Set([
  "period_id_ord", "period_id_trend", "period_id_of_cycle", "horizon",
  "period_id_sin", "period_id_cos", "period_id_sin2", "period_id_cos2",
  "period_id_quarter", "period_id_month", "month_of_year",
  "quarter_of_year", "is_quarter_start",
]);
const NON_FEATURES = new Set(["adversarial_weights"]);

// ── numeric helpers (match numpy semantics) ──────────────────────────────────
function mean(arr) {
  let s = 0;
  for (const x of arr) s += x;
  return s / arr.length;
}

// Population variance (np.var, ddof=0).
function popVar(arr) {
  const m = mean(arr);
  let s = 0;
  for (const x of arr) s += (x - m) * (x - m);
  return s / arr.length;
}

// pandas to_numpy(dtype=float): null/undefined -> NaN, else Number().
function toFloat(v) {
  if (v === null || v === undefined) return NaN;
  return Number(v);
}

// pandas to_numeric(errors="coerce"): non-parseable -> NaN.
function toNumeric(v) {
  if (typeof v === "number") return v;
  if (typeof v === "string" && v.trim() !== "" && !Number.isNaN(Number(v))) return Number(v);
  return NaN;
}

function isNumericColumn(rows, col) {
  return rows.every((r) => {
    const v = r[col];
    return v === null || v === undefined || typeof v === "number";
  });
}

// ── public surface ───────────────────────────────────────────────────────────

// |mean(a) - mean(b)| / pooled_std. Effect size; 0 = identical; does NOT
// saturate the way an adversarial AUC does. Drops NaNs; <5 valid on either side
// -> NaN; pooled_std == 0 -> 0.0.
export function standardizedMeanShift(a, b) {
  a = Array.from(a, Number).filter((v) => !Number.isNaN(v));
  b = Array.from(b, Number).filter((v) => !Number.isNaN(v));
  if (a.length < 5 || b.length < 5) return NaN;
  const sp = Math.sqrt((popVar(a) + popVar(b)) / 2.0);
  if (sp === 0.0) return 0.0;
  return Math.abs(mean(a) - mean(b)) / sp;
}

// Recent-vs-full drift diagnostic. Returns { ok, ... } or { ok: false, reason }.
// The val TARGET is never read — only shared numeric covariates participate.
export function driftDiagnostic(train, val, timeCol, opts = {}) {
  const recentPeriods = opts.recentPeriods ?? RECENT_PERIODS;
  const exclude = opts.exclude ?? null;

  const excluded = new Set([...WINDOW_FEATURES, ...NON_FEATURES]);
  if (exclude) for (const e of exclude) excluded.add(e);

  if (val == null || train == null) {
    return { ok: false, reason: "no val frame available" };
  }
  const trainCols = train.length ? Object.keys(train[0]) : [];
  if (!timeCol || !trainCols.includes(timeCol)) {
    return { ok: false, reason: `time_col '${timeCol}' not present in train frame` };
  }

  const valCols = new Set(val.length ? Object.keys(val[0]) : []);
  const common = trainCols.filter(
    (c) => valCols.has(c) && isNumericColumn(train, c) && !excluded.has(c),
  );
  if (common.length === 0) {
    return { ok: false, reason: "no shared numeric non-window features between train and val" };
  }

  const trRank = train.map((r) => toNumeric(r[timeCol]));
  const trRankValid = trRank.filter((v) => !Number.isNaN(v));
  if (trRankValid.length === 0) {
    return { ok: false, reason: "could not derive numeric rank for train rows" };
  }

  // Iterative max (NOT Math.max(...trRankValid)): spreading a large array into a
  // call overflows the argument stack (~125k elements → "Maximum call stack size
  // exceeded"). This is numerically identical for finite ranks.
  let rankMaxRaw = -Infinity;
  for (const v of trRankValid) if (v > rankMaxRaw) rankMaxRaw = v;
  const rankMax = Math.trunc(rankMaxRaw);
  const rankSpan = rankMax + 1;
  if (rankSpan < recentPeriods) {
    return { ok: false, reason: `train spans ${rankSpan} periods < RECENT_PERIODS=${recentPeriods}` };
  }
  const cutoff = rankMax - recentPeriods + 1;
  const recentMask = trRank.map((v) => (Number.isNaN(v) ? false : v >= cutoff));

  const rows = [];
  for (const c of common) {
    const aFull = train.map((r) => toFloat(r[c]));
    const aRecent = train.filter((_, i) => recentMask[i]).map((r) => toFloat(r[c]));
    const bVal = val.map((r) => toFloat(r[c]));
    const dFull = standardizedMeanShift(aFull, bVal);
    const dRecent = standardizedMeanShift(aRecent, bVal);
    if (Number.isNaN(dFull) || Number.isNaN(dRecent)) continue;
    rows.push([c, dFull, dRecent, dFull - dRecent]);
  }

  if (rows.length === 0) {
    return { ok: false, reason: "no feature yielded a valid distance (too many NaNs?)" };
  }

  const meanFull = mean(rows.map((r) => r[1]));
  const meanRecent = mean(rows.map((r) => r[2]));
  const meanImpr = mean(rows.map((r) => r[3]));
  // STRICT >: an improvement of exactly 0.05 must NOT count.
  const fracImproved = mean(
    rows.map((r) => (r[3] > DRIFT_IMPROVE_PER_FEATURE_THRESHOLD ? 1 : 0)),
  );
  const rel = meanFull > 0 ? meanImpr / meanFull : 0.0;

  return {
    ok: true,
    n_features_scanned: rows.length,
    rank_max: rankMax,
    recent_periods: recentPeriods,
    recent_cutoff_rank: cutoff,
    mean_dist_full: meanFull,
    mean_dist_recent: meanRecent,
    mean_improvement: meanImpr,
    rel,
    frac_improved: fracImproved,
    per_feature: rows.map(([c, dFull, dRecent]) => ({
      feature: c,
      dist_full: dFull,
      dist_recent: dRecent,
      improvement: dFull - dRecent,
    })),
  };
}

// Python f"{x:.3f}": round-half-to-even at 3 decimals.
function fmt3(x) {
  return x.toFixed(3);
}

// Map a drift diagnostic onto the CV scheme. Expanding is the default; sliding
// requires ALL three gates. Returns { scheme, reason, gates }.
export function decideScheme(diagnostic) {
  if (!diagnostic || !diagnostic.ok) {
    const reason = diagnostic && diagnostic.reason !== undefined ? diagnostic.reason : "unknown";
    return {
      scheme: "expanding",
      reason: `expanding (fallback - drift diagnostic could not run: ${reason})`,
      gates: { frac_ok: false, rel_ok: false, nfeat_ok: false },
    };
  }

  const rel = diagnostic.rel;
  const frac = diagnostic.frac_improved;
  const nFeat = Math.trunc(Number(diagnostic.n_features_scanned) || 0);
  const fracOk = frac >= DRIFT_FRAC_IMPROVED_THRESHOLD;
  const relOk = rel >= DRIFT_REL_THRESHOLD;
  const nfeatOk = nFeat >= MIN_FEATURES_FOR_SLIDING;
  const gates = { frac_ok: fracOk, rel_ok: relOk, nfeat_ok: nfeatOk };

  if (fracOk && relOk && nfeatOk) {
    const reason =
      `sliding (drift is recency-reducible AND evidence is broad enough: ` +
      `frac_improved=${fmt3(frac)} >= ${DRIFT_FRAC_IMPROVED_THRESHOLD} ` +
      `AND rel=${fmt3(rel)} >= ${DRIFT_REL_THRESHOLD} ` +
      `AND n_features=${nFeat} >= ${MIN_FEATURES_FOR_SLIDING})`;
    return { scheme: "sliding", reason, gates };
  }

  const fails = [];
  if (!nfeatOk) {
    fails.push(
      `insufficient features to assess drift breadth ` +
        `(n_features=${nFeat} < ${MIN_FEATURES_FOR_SLIDING})`,
    );
  }
  if (!fracOk) {
    fails.push(`frac_improved=${fmt3(frac)} < ${DRIFT_FRAC_IMPROVED_THRESHOLD} (primary gate)`);
  }
  if (!relOk) {
    fails.push(`rel=${fmt3(rel)} < ${DRIFT_REL_THRESHOLD} (secondary gate)`);
  }
  const reason =
    "expanding (conservative default; sliding not affirmatively justified - " +
    fails.join("; ") +
    ")";
  return { scheme: "expanding", reason, gates };
}
